### DISCUSSION TITLE
Fetching from API in getStaticProps returning HTML on vercel

### DISCUSSION THREAD

**Discussion by dandesdev** [vercel/next.js#23709] [Help]
Recently Ive posted another error I was facing when deploying my project to Vercel. I was able to find a weird solution to that wich I posted already on that topic. But now Im getting another weird thing. This is my code:

```
const envPath = process.env.NODE_ENV === "development" ? 
"http://localhost:3000"
: `https://${process.env.NEXT_PUBLIC_VERCEL_URL}`;

const bookPath = "/api/book";

export async function getStaticProps(context) {

    const book = await fetch(envPath + bookPath).then((res) => res.json()) // <-- This was outputtingan error
    return {
        props: {
            book
        }
    }
}
```

When I had the code like that I was getting that commom error 'Invalid token < in JSON at position 0...'. My api route has only a test function for now:

```
export default function getApi(req, res) {
    res.json("Hello world");
}
```

so I changed the code in my index js to output the response as a text to see what was going wrong, like this:

`const fetchTest = await fetch(envPath + bookPath).then((res) => res.text()).then((text) => console.log(text));`

And then I saw, in the vercel log in build step I would get an entire HTML page as this text variable, starting with:

`<!DOCTYPE html><html lang="en"><head><meta charSet="utf-8"/><title>Login – Vercel</title><meta name="viewport" cont...`

I even searched for "Hello world" in the text and found nothing. I dont understand why the fetch promise is resolving into a HTML code in Vercel server instead of a json as it happens in my localhost.
Anyone knows what could be going on?

**Comment by wwwdepot** [comment]
Per [Data Fetching](https://nextjs.org/docs/basic-features/data-fetching)

```
Note: You should not use fetch() to call an API route in getStaticProps. Instead, directly import the logic used inside your API route. You may need to slightly refactor your code for this approach.

Fetching from an external API is fine!
```
